go/database/mpt/io/parallel_visit.go

```
@@ -0,0 +1,429 @@
+type NodeSource interface {
+	Get(mpt.NodeId) (mpt.Node, error)
+	Close() error
+}
+
+// visitAll visits all nodes in the trie rooted at the given node.
+// This method accesses nodes side to another infrastructure that may be already accessing the nodes.
+// The nodes are read multithreading, but provided via the visitor as they appear DFS in the trie.
+func visitAll(
+	sourceFactory NodeSourceFactory,
+	root mpt.NodeId,
+	visitor mpt.NodeVisitor,
+	cutAtAccounts bool,
+) error {
+
+	// The idea is to have workers processing a common queue of needed
+	// nodes sorted by their position in the depth-first traversal of the
+	// trie. The workers will fetch the nodes and put them into a shared
+	// map of nodes. The main thread will consume the nodes from the map
+	// and visit them.
+
+	type request struct {
+		position *position
+		id       mpt.NodeId
+	}
+
+	requestsMutex := sync.Mutex{}
+	requests := heap.New(func(a, b request) int {
+		return b.position.compare(a.position)
+	})
+
+	type response struct {
+		node mpt.Node
+		err  error
+	}
+	responsesMutex := sync.Mutex{}
+	responsesCond := sync.NewCond(&responsesMutex)
+	responses := map[mpt.NodeId]response{}
+
+	done := atomic.Bool{}
+	defer done.Store(true)
+
+	requests.Add(request{nil, root})
+
+	const NumWorker = 16
+
+	// open all sources for the number of workers and track possible errors
+	sources := make([]NodeSource, 0, NumWorker)
+	openingErrors := make([]error, 0, NumWorker)
+	for i := 0; i < NumWorker; i++ {
+		source, err := sourceFactory.Open()
+		sources = append(sources, source)
+		openingErrors = append(openingErrors, err)
+	}
+
+	defer func() {
+		// only log errors when closing the sources
+		// as data are opened parallel to the main program
+		// and thus the error should not invalidate the result
+		for _, source := range sources {
+			if source != nil {
+				if err := source.Close(); err != nil {
+					log.Printf("failed to close source: %v", err)
```

These errors get lost; they should be reported as part of the result of this call.

---

reverted back to open/close in the go-routine and sending errors via a channel